### ISSUE TITLE
using workerpool in production

### ISSUE THREAD

**Issue by JohanJermey** [oze4/workerpoolxt#12]
Hey,

Can we use it on production? 
if so did you run some go-profiling on bulk of workers? 

Thanks

**Comment by oze4** [comment]
I'm not sure I would use this in production as of yet.

I would recommend doing something like this for now (until I can update this repo)

```golang
package job

// Job knows how to run on it's own goroutine, as well as how to be submitted
// to a `*workerpool.WorkerPool`
type Job interface {
	Name() string
	Run()
}

// NewJob creates a new job
func NewJob(name string, resultChan chan Result, todo func() (interface{}, error)) Job {
	return &job{name: name, result: resultChan, todo: todo}
}

// job implements Job
type job struct {
	name   string
	result chan Result
	todo   func() (interface{}, error)
}

// Name returns job name
func (j *job) Name() string {
	return j.name
}

// Run invokes the job
func (j *job) Run() {
	res, err := j.todo()
	j.result <- Result{jobName: j.name, data: res, err: err}
}

// Result holds job result
type Result struct {
	jobName string
	data    interface{}
	err     error
}

// JobName returns the Job name
func (r *Result) JobName() string {
	return r.jobName
}

// Err returns Job error
func (r *Result) Err() error {
	return r.err
}

// Data returns Job data
func (r *Result) Data() interface{} {
	return r.data
}
```

There was a "[bug](https://github.com/gammazero/workerpool/issues/40#issue-726019885)" discovered where even though a job may timeout via context, and appear to be cancelled, it will still run behind the scenes.  **There is no way to stop a goroutine from running once you've started it (from outside of it)**.  Therefore, even though a job may appear to have timed out, it will still consume resources behind the scenes.

I reached out to the author of `workerpool` (which is what we wrap around), and he had [this](https://github.com/gammazero/workerpool/issues/40#issuecomment-713709976) to say.  From what I make of it, the way Go works, plus the way `workerpool` is written, makes me believe `workerpoolxt` was an "over-abstraction" (meaning, `workerpool`'s abstractions are currently sufficient if used correctly).

Unfortunately, stopping a goroutine from outside of it is rather tricky.

**Comment by oze4** [comment]
[Another](https://github.com/oze4/workerpoolxt/issues/7#issuecomment-713731079) useful link...

**Comment by JohanJermey** [comment]
@oze4  - you write "I'm not sure I would use this in production as of yet" , becouse of the cancelation(or not ) of the spawned running go routine or there is other things, if just this we can use it? 

btw,
I really like the idea of the `workerpoolxt`, 
I read the reference you provided, If you want to ask me, I will remove the tricky parts and put only the **basics** at first, then add the feature's slowly after comprehensive tests .

**Comment by oze4** [comment]
If you'd like to use the basics, I would recommend just using `workerpool` directly for now.

The issue with cancelling a goroutine from the outside is the only issue I have found. It just gives the impression the job times out or was cancelled, but the job will continue to run.

**Comment by JohanJermey** [comment]
@oze4  but isnt it the same issue with the workerpool package?  if I use the workerpool directly I will not have this issue, I doubts but pls correct me if im wrong..

**Comment by oze4** [comment]
The "issue" is just how go works. You'd have this same issue regardless of what you're using. But to answer your question, this issue IS NOT in workerpool because workerpool doesn't try to abstract jobs like I was trying to do.

**Comment by oze4** [comment]
Basically once you give a func to workerpool you can't cancel it mid-execution. If you aren't trying to cancel jobs mid-execution then you won't have this issue.

**Comment by JohanJermey** [comment]
After digging a bit I saw this 
https://github.com/gammazero/workerpool/issues/40#issuecomment-714178180
isnt it the solution for the problem? if so, do you plan to adopt it?

Many thanks

**Comment by JohanJermey** [comment]
@oze4  - Thanks for all the responses and kind answers.

 One thing I definitely missing here.

see https://play.golang.org/p/FKmLyjaPbav

The jobs are running with timeouts and loops and get killed after the contexts timeout, so what am I missing here?  you mention that the job will proceed to run

**Comment by oze4** [comment]
See [here](https://play.golang.org/p/ipfoEPL1cg1). Your demo uses the same context for all jobs, which will work. What if you wanted to pass a different context to each job, though? 

[This](https://play.golang.org/p/xr8IJ1uM4tV) is the same example as the one above, but each job is using the same context. This works without issue - the problem is supplying a context with each job.  Providing a different context per job is more flexible, but if we want to get rid of that option, which means less flexibility, then that solves our issue.  I just don't like the idea of all jobs timing out just because one job timed out (or was cancelled, etc..).

I do plan on adopting the new idea eventually.  With that being said, there is no way around this issue.

**Comment by oze4** [comment]
https://github.com/oze4/job

I think that is a good start for proof-of-concept.

**Comment by JohanJermey** [comment]
Thanks @oze4  - I think using one contexts(timeout) for all is also bring a **lot of benefit** for scenario that you need to provide a response to some api/server/etc and you limit all the worker pull for that time. e.g. provide a limit of 1 min and kill all the process running after that.  what do you think ?

**Comment by oze4** [comment]
I think that change is the easiest to make. I plan on making these changes when I find some time.

**Comment by oze4** [comment]
https://github.com/chebyrash/promise is another good solution... works like Node promises.

**Comment by JohanJermey** [comment]
Ok nice. but I think it will not kill the process either...one thing that make my a bit confused is how when you use the same contexts you can kill the go routines and when you use different contexts you cannot, cloud you please explain?

**Comment by oze4** [comment]
That Promise library does not have the same problem.. 

In order to kill a goroutine, it needs a mechanism **inside** of it to escape. All `workerpool` does is accept a `func()`... and all `workerpoolxt` does is wrap around that func via `Job.Task`.

### TLDR

At a high level, this code summarizes the issue: https://play.golang.org/p/Yeqeb3g3PSV

```golang
package main

import (
	"fmt"
	"time"
)

func main() {
	takeFuncAsParam(func() {
		fmt.Println("from func")
	})

	time.Sleep(time.Second * 5)
}

func takeFuncAsParam(f func()) {
	go f() // <- how do you stop this from running?
	fmt.Println("[takeFuncAsParam] Done.")
}
```

### Detailed Answer

Actually, after testing some more, it doesn't matter if you provide context at the job or pool level - it will still run regardless of whether your context has been cancelled or not.  This is because the code you provide inside of the func() does not have a way to "exit".

[Please see this Playground](https://play.golang.org/p/r15QMmMGika) which demonstrates how it doesn't matter if you use context at the pool or job level (code for the go playground is also below). ***Please read the comments on line 41***

#### How `workerpool` works

```golang
func main() {
    wp := workerpool.New(10) // Max 10 goroutines run at a time

    wp.Submit(func() {
        fmt.Println("first func running")
        time.Sleep(time.Second * 2)
        // Do whatever.....
    })

    wp.Submit(func() {
        fmt.Println("second func running")
        time.Sleep(time.Second * 3)
        // Do whatever.....
    })

    wp.Submit(func() {
        fmt.Println("third func running")
        time.Sleep(time.Second * 4)
        // Do whatever.....
    })

    // Must call StopWait so the main thread waits on all spawned goroutines
    wp.StopWait()

    // Continue doing whatever.....
}
```

Again, all `workerpoolxt` does is wrap around the func() that gets passed into `workerpool`

```golang
// using the code above, all we do is wrap around the func..
func() {
        fmt.Println("third func running")
        time.Sleep(time.Second * 4)
        // Do whatever.....
    }
```

This is what `workerpoolxt` does when you give us a Job:

```golang
func (r *runner) wrap(ctx context.Context, job *job) func() {
	return func() {
		job.result = make(chan jobResult)

		go job.Run()

		select {
		case res := <-job.result:
			r.result <- res
		case <-ctx.Done():
			fmt.Printf("Job '%s' should stop here\n", job.Name)
			r.result <- jobResult{name: job.Name, Error: ctx.Err()}
		}
	}
}
```

Notice we call `go job.Run()`... Which just calls your task, and returns the result on a channel:

```golang
func (j *job) Run() {
	result := j.Task()
	result.name = j.Name
	j.result <- result
}
```

If you notice, inside of the func `wrap(..)` we wait on a response from either the Result channel or the context - whichever happens first we will take and aggregate the Result.  But just because the context was cancelled does not kill the goroutine that was started by `workerpool`.  It has no idea you're even using context - how would it know to end?


```golang
package main

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/gammazero/workerpool"
)

func main() {
	myctx, cancelmyctx := context.WithTimeout(context.Background(), time.Duration(time.Second*3))
	defer cancelmyctx()

	runner := newRunner(myctx, 10)

	runner.do(job{
		Name: "a",
		Task: func() jobResult {
			for i := 0; i < 100000000; i++ {
				time.Sleep(time.Second * 1)
				msg := "job 'a' running..\n"
				fmt.Printf(msg)
			}
			return jobResult{Data: "from a"}
		},
	})

	runner.do(job{
		Name: "b",
		Task: func() jobResult {
			time.Sleep(time.Duration(time.Second * 6))
			return jobResult{Data: "from b"}
		},
	})

	results := runner.getjobResults()
	fmt.Println(results)

	// Pretend you are doing other stuff here...
	// ***************************************************
	// As long as this main thread is open, the goroutine spawned by workerpool
	// will continue to run
	time.Sleep(time.Second * 8)
}

type runner struct {
	*workerpool.WorkerPool
	ctx     context.Context
	kill    chan bool
	once    sync.Once
	result  chan jobResult
	results []jobResult
}

func newRunner(ctx context.Context, numRunners int) *runner {
	r := &runner{
		WorkerPool: workerpool.New(numRunners),
		ctx:        ctx,
		kill:       make(chan bool),
		result:     make(chan jobResult),
	}

	go r.processResults()

	return r
}

func (r *runner) processResults() {
	for {
		select {
		case res, ok := <-r.result:
			if !ok {
				goto Done
			}
			r.results = append(r.results, res)
		}
	}
Done:
	<-r.kill
}

func (r *runner) do(j job) {
	r.Submit(r.wrap(r.ctx, &j))
}

func (r *runner) getjobResults() []jobResult {
	r.once.Do(func() {
		r.StopWait()
		close(r.result)
		r.kill <- true
	})
	return r.results
}

func (r *runner) wrap(ctx context.Context, job *job) func() {
	return func() {
		job.result = make(chan jobResult)

		go job.Run()

		select {
		case res := <-job.result:
			r.result <- res
		case <-ctx.Done():
			fmt.Printf("Job '%s' should stop here\n", job.Name)
			r.result <- jobResult{name: job.Name, Error: ctx.Err()}
		}
	}
}

type job struct {
	Name   string
	Task   func() jobResult
	result chan jobResult
}

func (j *job) Run() {
	result := j.Task()
	result.name = j.Name
	j.result <- result
}

type jobResult struct {
	name  string
	Error error
	Data  interface{}
}
```

**Comment by BredSt** [comment]
WOW I cannot thank you enough on the detailed answer .now it's much more clear, I miss the main tread.
why you said the the promise lib doesn't have this issue? did you try it? if I could put a bet on it, I put my money that it have the  the **exact same issue**.

**Comment by JohanJermey** [comment]
WOW I cannot thank you enough on the detailed answer .now it's much more clearer to me, I miss the main tread. btw, why you said the the promise lib doesn't have this issue? did you try it? if I could put a bet on it, I put my money that it have the the exact same issue.

**Comment by oze4** [comment]
Well, I can't say for certain there aren't any issues with it because I haven't tested but I assume they thought of this type of thing. 

Besides that, the logic and architecture of Promise are fundamentally different than that of workerpool. 

Although the end result may appear to be similar, under the hood things are different. It doesn't look like Promise is concurrency limited - its just syntactical sugar much like Promises are in Node. There is no concept of a job queue or worker threads in Promise - it allows you to abstract that sort of thing.

**Comment by oze4** [comment]
@JohanJermey 

I just wanted to put this here since I believe it may help. I did not fully understand until I wrote this package. 

Workerpool is a fine piece of art imo. 

Let's say you have 2,000 resource intensive jobs to run. Sure, you could start all 2,000 jobs easily on their own goroutines. Bam. Concurrency. 

Then your computer slows to a crawl and all 2k jobs take even longer. 

Workerpool will limit the number of "active" jobs while queuing anything over your max worker count. It does this flawlessly. So much so that it is easy to overlook the intricacies and details of what it is really doing. It's beautiful lol

Why do I mention this? Because a workerpool my not be what you're looking for to begin with.

**Comment by JohanJermey** [comment]
HI Again,

I think that the promise have the same issue even it's different implementation.  

https://github.com/chebyrash/promise/issues/23
what do you think ?

**Comment by JohanJermey** [comment]
@oze4  - I've a direction, lets say that in the constructor user will provide the overall timeout (lets say 10 sec for all the jobs) 
and inside the jobs user will need to call also to another function inside his job. and this function will check if the job pass this limit time that defined inside the constructor and if yes terminte it from inside the job function, what do you think
?

**Comment by JohanJermey** [comment]
@oze4 - any comments :)

**Comment by oze4** [comment]
@JohanJermey apologies, I haven't had a chance to look at it yet. I'll try to take a look later today

**Comment by oze4** [comment]
@JohanJermey I finally had some time to look at this but both of those links (the code, not the website) error out?

**Comment by JohanJermey** [comment]
@oze4 - I think you need to click on the import https://play.golang.org/p/eV_8GFd3SXi this working for me

**Comment by oze4** [comment]
That one works. But doesn't it just show the same issue? It looks like job A keeps running even after context has been cancelled?
